Pass a string to NDArray(name=...) and raise NotImplemented on named models in SMC

basnijholt opened the pull request:
Currently, `name` is a `Model`.


**Thank your for opening a PR!**

Depending on what your PR does, here are a few things you might want to address in the description:
+ [x] what are the (breaking) changes that this PR makes? → see thread & release notes
+ [x] important background, or details about the implementation → see thread
+ [x] are the changes—especially new features—covered by tests and docstrings?
+ [ ] <s>consider adding/updating relevant example notebooks</s>
+ [x] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

basnijholt committed:
pass a string to NDArray(name=...)

Currently, `name` is a `Model`.

MarcoGorelli commented:
Nice, thanks @basnijholt ! Is it possible to add a little test somewhere, or to change an existing test so that this is checked?

basnijholt commented:
@MarcoGorelli, there is no public API that exposes this directly (AFAIK).

OTOH, these types of errors are easily caught by adding type hints and using `mypy`.

twiecki commented:
A test would be good regardless, apparently this code-path wasn't tested before.

basnijholt committed:
Add tests related to named models

A test was added to asserts that strace.name is a string (this was not the case before #4365).

Non-empty model names are actually not supported (again, see #4365) so attempting to SMC-sample a named model will now raise a NotImplementedError.

basnijholt commented:
@twiecki, @MarcoGorelli, I have added a test.

MarcoGorelli commented on pymc3/tests/test_smc.py:
```diff
@@ -189,3 +189,8 @@ def test_repr_latex(self):
+    def test_name_is_string_type(self):
+        with self.SMABC_potential:
+            trace = pm.sample_smc(draws=10, kernel="ABC")
+            assert isinstance(trace._straces[0].name, str)
```
could this line be added to the end of a previous test (e.g. `test_potential`)? The test suite is already pretty long

Also, it is possible to check the actual names, rather than just making assertions about the types? E.g. something like
```python
assert [i.name for i in trace._straces] == ['foo', 'bar']
```

michaelosthege replied:
Thanks @MarcoGorelli for requesting this change. Turns out named models aren't actually supported. (see main comment)

MarcoGorelli reviewed (changes requested):
Thanks for adding a test!

Can you also add a note to the release notes?

michaelosthege committed:
Mention SMC/named models related PR #4365

michaelosthege commented:
While making the test more precise, I found that the implementation does not actually support having a model name.

The reason is that the model name becomes a prefix to the variable name, which is also used as a kwarg in the line `sim_data = self.function(**func_parameters)`. The function that is called is user-provided to the `Simulator` and does now have the prefix in the kwarg.

So these prefixes would have to be eliminated from the kwargs, but then the corner case of having nested models is hard to deal with. (_yes, that's possible and no, nobody does that because it's very fragile_)
With the upcoming changes towards `RandomVariable` `Op`s, we might get rid of the `Model` context entirely. Therefore I think we should not invest precious dev time into fixing this fragility.

I'm going to insert this before the `NDArray` creation:
```python
if self.model.name:
    raise NotImplementedError(
        "The SMC implementation currently does not support named models. "
        "See https://github.com/pymc-devs/pymc3/pull/4365."
    )
```

michaelosthege commented:
I just rebased to fix the README conflict. @aloctavodia if you approve, I think we can merge this.
